pkg/accountability/simpleacc/internal/predecisions/predecisions.go

```diff
@@ -0,0 +1,171 @@
+package predecisions
+
+import (
+	"reflect"
+
+	"github.com/filecoin-project/mir/pkg/accountability/simpleacc/common"
+	incommon "github.com/filecoin-project/mir/pkg/accountability/simpleacc/internal/common"
+	"github.com/filecoin-project/mir/pkg/accountability/simpleacc/internal/poms"
+	"github.com/filecoin-project/mir/pkg/dsl"
+	"github.com/filecoin-project/mir/pkg/logging"
+	accpbdsl "github.com/filecoin-project/mir/pkg/pb/accountabilitypb/dsl"
+	accpbmsgs "github.com/filecoin-project/mir/pkg/pb/accountabilitypb/msgs"
+	accpbtypes "github.com/filecoin-project/mir/pkg/pb/accountabilitypb/types"
+	cryptopbdsl "github.com/filecoin-project/mir/pkg/pb/cryptopb/dsl"
+	cryptopbtypes "github.com/filecoin-project/mir/pkg/pb/cryptopb/types"
+	transportpbdsl "github.com/filecoin-project/mir/pkg/pb/transportpb/dsl"
+	t "github.com/filecoin-project/mir/pkg/types"
+	"github.com/filecoin-project/mir/pkg/util/maputil"
+	"github.com/filecoin-project/mir/pkg/util/membutil"
+)
+
+// IncludePredecisions implements the broadcast and treatment of predecisions.
+// The broadcast and treatment of certificates is
+// externalized to the ../certificates package so that a system can decide whether to instantiate the accountability
+// module with full certificates or with the optimistic light certificates.
+// The broadcast and treatment of PoMs is also externalized to the ../poms package.
+func IncludePredecisions(
+	m dsl.Module,
+	mc *common.ModuleConfig,
+	params *common.ModuleParams,
+	state *incommon.State,
+	logger logging.Logger,
+) {
+
+	accpbdsl.UponPredecided(m, func(data []byte) error {
+
+		if state.Predecided {
+			logger.Log(logging.LevelWarn, "Received Predecided message while already Predecided")
+			return nil
+		}
+
+		state.Predecided = true
+
+		// Sign predecision attaching mc.Self to prevent replay attacks
+		cryptopbdsl.SignRequest(m, mc.Crypto, &cryptopbtypes.SignedData{Data: [][]byte{data, []byte(mc.Self)}}, &signRequest{data: data})
+
+		return nil
+	})
+
+	cryptopbdsl.UponSignResult(m, func(signature []byte, sr *signRequest) error {
+		state.SignedPredecision = &accpbtypes.SignedPredecision{
+			Predecision: sr.data,
+			Signature:   signature,
+		}
+
+		// Broadcast signed predecision to all participants (including oneself)
+		transportpbdsl.SendMessage(m, mc.Net, accpbmsgs.SignedPredecision(mc.Self, sr.data, signature), maputil.GetKeys(params.Membership.Nodes))
```

Comment:
You might need to send this message periodically.

Reply:
Done, for this one and also for a new one that is required now that we do not have just []byte as the decision (i.e. leader, aborted, etc.). We can omit the new message with a deserialize function, but I feel like this is more general and allows for hashing of the serialized version. What do you think?